render/jpeg2k/jpeg2k_encoder_cgo.go

```diff
@@ -0,0 +1,145 @@
+/*
+ *
+ * An implementation example of a custom encoder for handling JPEG2000 image format.
+ *
+ */
+
+package jpeg2k
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/unidoc/unipdf/v3/core"
+	"gopkg.in/gographics/imagick.v2/imagick"
+)
+
+// Custom encoder declaration that implements StreamEncoder interface.
+type CustomJPXEncoder struct{}
+
+// NewCustomJPXEncoder returns a new instance of CustomJPXEncoder.
+func NewCustomJPXEncoder() *CustomJPXEncoder {
+	return &CustomJPXEncoder{}
+}
+
+// GetFilterName returns the name of the encoding filter.
+func (enc *CustomJPXEncoder) GetFilterName() string {
+	return "CustomJPXEncoder"
+}
+
+// MakeDecodeParams makes a new instance of an encoding dictionary based on
+// the current encoder settings.
+func (enc *CustomJPXEncoder) MakeDecodeParams() core.PdfObject {
+	return nil
+}
+
+// MakeStreamDict makes a new instance of an encoding dictionary for a stream object.
+func (enc *CustomJPXEncoder) MakeStreamDict() *core.PdfObjectDictionary {
+	dict := core.MakeDict()
+
+	dict.Set("Filter", core.MakeName(enc.GetFilterName()))
+
+	return dict
+}
+
+// UpdateParams updates the parameter values of the encoder.
+func (enc *CustomJPXEncoder) UpdateParams(params *core.PdfObjectDictionary) {
+}
+
+// DecodeBytes decodes a slice of JPX encoded bytes and returns the result.
+func (enc *CustomJPXEncoder) DecodeBytes(encoded []byte) ([]byte, error) {
+	imagick.Initialize()
+	defer imagick.Terminate()
+
+	mw := imagick.NewMagickWand()
+	defer mw.Destroy()
+
+	if err := mw.ReadImageBlob(encoded); err != nil {
+		return nil, err
+	}
+
+	// fmt.Println(mw.IdentifyImage())
+
+	imgWidth := mw.GetImageWidth()
+	imgHeight := mw.GetImageHeight()
+	imgDepth := mw.GetImageDepth()
+	cs := mw.GetImageColorspace()
+
+	fmt.Printf("Color space %v - %d bit\n", cs, imgDepth)
+
+	imw := mw.NewPixelIterator()
+	defer imw.Destroy()
+
+	colorComponent := 3
+
+	switch cs {
+	case imagick.COLORSPACE_GRAY:
+		colorComponent = 1
+	case imagick.COLORSPACE_CMYK:
+		colorComponent = 4
+	}
+
+	var decoded = make([]byte, imgWidth*imgHeight*uint(colorComponent)*imgDepth/8)
+	index := 0
+
+	for y := 0; y < int(imgHeight); y++ {
+		pmw := imw.GetNextIteratorRow()
+
+		for x := 0; x < int(imgWidth); x++ {
+			if cs == imagick.COLORSPACE_GRAY {
+				r := int(pmw[x].GetRed() * 255)
```

Comment:
I think that should be rather `GetBlack` instead of `GetRed`

Reply:
After further testing, using `GetBlack` on grayscale image would render a black box instead of the actual image.

It doesn't have to `GetRed` though, I've tested using `GetGreen` and `GetBlue` and it correctly render the grayscale image.